### PR TITLE
Decoupling

### PR DESCRIPTION
**Pull request by BenoitJeanson**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
no

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Refactoring. Main goals:
- Remove cyclic dependencies of classes in the model package:
- Organize the classes of package `model` into subclasses (`nodes`, `blocks`, `cells`, `graphs`)

Results in strictly oriented dependencies of the `model `subpackages:
- `graphs `depends on `nodes `and `cells`
- `cells `depends on `nodes `and `blocks`
- `blocks `depends on `nodes`

`Graph `becomes the central routing class with new features. Mainly:
- `VoltageLevelGraph getVoltageLevelGraph (Node, node)` to replace the `Node::getVoltageLevelGraph`. This is based on `Map<Node,VoltageLevel> nodeToVlGraph` that is shared by all the inherited classes of Graph (`ZoneGraph`, `SubstationGraph`, `VoltageLEvelGraph`, ... the constructor needs the `BaseGraph parentGraph` to ensure unicity of this map).
- no longer need of the boolean `forVoltageLevelOnly` as a `parentGraph` is given when it should be `false`
- `Optional<Cell> getCell(Node node)` to repalce `Node::getCell`

To ensure the consistency of the `Graph `creation, `graphs.NodeFactory` gathers all the node creation methods and ensure the nodes are added to the `Graph` it belongs to (most often to the `VoltageLevelGraph`, but for `Middle2WTNode` and `Middle3WTNode ` that are outside of a `VoltageLevelGraph`, the connection to the appropriate `BaseGraph`)

**What is the new behavior (if this is a feature change)?**
For `GraphBuilder` implementations :
- Change in the building of a `Graph` (constructors evolve)
- Recommendation to use `graph.NodeFactory` instead of `Node` constructors

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [x] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:
Next step in progress is the decoupling of `layout` and `model` as `model` depends on `layout.LayoutParameters` (and `layout` depends on `model`).
Then, working on #308. This will probably imply a new breaking change in `Graph` building: it sounds wise to wait for that feature before issuing a new release.

### CODE REVIEW

**Review by flo-dup — CHANGES_REQUESTED** [review]
Nice refactor! Some feedback which I'd like to talk with you - I could take care of it afterwards.

File: single-line-diagram-core/src/main/java/com/powsybl/sld/layout/PositionVoltageLevelLayout.java [hunk 1]
```diff
@@ -86,17 +88,27 @@ private void adaptPaddingToSnakeLines(LayoutParameters layoutParam) {
...
                 .filter(cell -> cell.getType() == Cell.CellType.EXTERN
                         || cell.getType() == Cell.CellType.INTERN)
-                .forEach(cell -> cell.calculateCoord(graph, layoutParam));
+                .map(BusCell.class::cast)
+                .forEach(cell -> {
+                    LayoutContext layoutContext = LayoutContext.create(graph.getFirstBusY(), graph.getLastBusY(layoutParam), graph.getExternCellHeight(cell.getDirection()), cell.getDirection());
+                    if (cell.getType() == Cell.CellType.INTERN) {
+                        layoutContext.setInternCell(true);
+                        layoutContext.setFlat(((InternCell) cell).getShape() == Shape.FLAT);
+                        layoutContext.setUnileg(((InternCell) cell).getShape() == Shape.UNILEG);
```
[the remark this answers is not in the document]
Comment (BenoitJeanson) [reply single-line-diagram-core/src/main/java/com/powsybl/sld/layout/PositionVoltageLevelLayout.java:104]: layoutContext will be refactor to ensure decoupling of model and layout subpackage...

File: single-line-diagram-core/src/main/java/com/powsybl/sld/builders/GraphBuilder.java [hunk 2]
```diff
@@ -8,19 +8,21 @@
...
  * @author Franck Lecuyer <franck.lecuyer at rte-france.com>
  */
 public interface GraphBuilder {
 
     VoltageLevelGraph buildVoltageLevelGraph(String id,
-                                             boolean forVoltageLevelDiagram);
+                                             Graph parentGraph);
 
-    SubstationGraph buildSubstationGraph(String id);
+    VoltageLevelGraph buildOrphanVoltageLevelGraph(String id);
```
[the remark this answers is not in the document]
Comment (BenoitJeanson) [reply single-line-diagram-core/src/main/java/com/powsybl/sld/builders/GraphBuilder.java:21]: an OrphanVL has no parent (poor of him)... ok with appropriate javadoc

File: single-line-diagram-core/src/main/java/com/powsybl/sld/layout/AbstractLayout.java [hunk 3]
```diff
@@ -50,12 +53,13 @@ protected void manageSnakeLines(BaseGraph graph, LayoutParameters layoutParamete
...
 
-    protected static BusCell.Direction getNodeDirection(Node node, int nb) {
+    protected Direction getNodeDirection(Node node, int nb) {
         if (node.getType() != Node.NodeType.FEEDER) {
             throw new PowsyblException("Node " + nb + " is not a feeder node");
         }
-        BusCell.Direction dNode = node.getCell() != null ? ((ExternCell) node.getCell()).getDirection() : BusCell.Direction.TOP;
-        if (dNode != BusCell.Direction.TOP && dNode != BusCell.Direction.BOTTOM) {
+        Optional<Cell> oCell = getGraph().getCell(node);
+        Direction dNode = oCell.isPresent() ? ((ExternCell) oCell.get()).getDirection() : Direction.TOP;
```
[the remark this answers is not in the document]
Comment (BenoitJeanson) [reply single-line-diagram-core/src/main/java/com/powsybl/sld/layout/AbstractLayout.java:61]: ok

File: single-line-diagram-core/src/main/java/com/powsybl/sld/model/graphs/VoltageLevelGraph.java [hunk 4]
```diff
@@ -69,15 +73,20 @@
...
+
+    public static VoltageLevelGraph create(VoltageLevelInfos voltageLevelInfos, Graph parentGraph) {
+        return new VoltageLevelGraph(voltageLevelInfos, parentGraph);
     }
 
-    public static VoltageLevelGraph create(VoltageLevelInfos voltageLevelInfos, boolean forVoltageLevelDiagram) {
-        return new VoltageLevelGraph(voltageLevelInfos, forVoltageLevelDiagram);
+    public static VoltageLevelGraph createForVoltageLevelGraphDiagram(VoltageLevelInfos voltageLevelInfos) {
+        return new VoltageLevelGraph(voltageLevelInfos, null);
     }
```
[the remark this answers is not in the document]
Comment (BenoitJeanson) [reply single-line-diagram-core/src/main/java/com/powsybl/sld/model/graphs/VoltageLevelGraph.java:90]: sure

**Comment by BenoitJeanson** [comment]
All your comments sound relevant. I am happy if you implement the suggested changes.

**Pull request merged**